### ISSUE TITLE
Wrong handling of file mode flags in int 21h function

### ISSUE THREAD

**Issue by DGhost001** [dosbox-staging/dosbox-staging#1549]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

All Versions are affected, as this Bug is also in the original dosBox source

### What Operating System are you using?

Linux x86_64

### If Other OS, please describe

Ubuntu

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

When using non-trivial open flags for a file handle in the dos 21h interrupt handler, dosbox shows different behaviour than a real dos system.

For example, Turbo Assembler uses the flags 0x4a to check if a obj file already exists.
The problem here is, that in the flags the "reserved" bit is also set.

The original DOS ignores this reserved bit at all.

Dosbox does report FILE_ACCESS_FAILED, when encountering the set flag. (No matter if the file exists or not) 

The Bug is in the dos_loclal.cpp, where it checks the flags is also includes the reserved one:

```c
bool localDrive::FileOpen(DOS_File **file, char *name, Bit32u flags)
{
	const char *type = nullptr;
	switch (flags&0xf) {
	case OPEN_READ:        type = "rb" ; break;
	case OPEN_WRITE:       type = "rb+"; break;
	case OPEN_READWRITE:   type = "rb+"; break;
	case OPEN_READ_NO_MOD: type = "rb" ; break; //No modification of dates. LORD4.07 uses this
	default:
		DOS_SetError(DOSERR_ACCESS_CODE_INVALID);
		return false;
	}
[...]
```
It should only check for the actual accessflags like so:

```c
bool localDrive::FileOpen(DOS_File **file, char *name, Bit32u flags)
{
	const char *type = nullptr;
	switch (flags&0x7) {
	case OPEN_READ:        type = "rb" ; break;
	case OPEN_WRITE:       type = "rb+"; break;
	case OPEN_READWRITE:   type = "rb+"; break;
	case OPEN_READ_NO_MOD: type = "rb" ; break; //No modification of dates. LORD4.07 uses this
	default:
		DOS_SetError(DOSERR_ACCESS_CODE_INVALID);
		return false;
	}
[...]
```

### Steps to reproduce the behaviour.

Explain how to reproduce
1.  run tasm to compile a asm file
2. run tasm on the same file again --> Couldn't write object file error

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by Wengier** [comment]
While the current FileOpen function can certainly be improved, I am not yet able to replicate the said problem in my own TASM installation, which appeared to work in this case:

![image](https://user-images.githubusercontent.com/8216923/151710922-94f17f08-813c-46be-8a4a-edda0923466f.png)

Tried in both Windows and Linux. It will be great if you can post a test package showing the said issue.

**Comment by DGhost001** [comment]
Ok, after playing around a bit more, the command line arguments to tasm are important...

Here is a Screenshot triggering the Bug:

![Screenshot_20220130_193421](https://user-images.githubusercontent.com/3473767/151712652-0dff51d0-91cf-436a-b2fe-33567c2dd3c7.png)


Anyway will write a little Testprogram to demonstrate the bug.

**Comment by grapeli** [comment]
@DGhost001 
If you decide not to add extensions to the files, the operation will be successful.
`tasm /o dither,dither`

The boraland manual does not use file extensions in the examples. (they are added automatically).
http://www.bitsavers.org/pdf/borland/turbo_assembler/Turbo_Assembler_4.1_for_OS2_Users_Guide_1994.pdf

On the other hand, under Freedos, it works fine in all variants.

**Comment by DGhost001** [comment]
I Testet with DOS 6.22 and 3.30, both also work fine with the file extensions.

Unfortunately  I'm not calling tasm directly and TurboC for example adds the file extensions on its own, when having ASM files in the project.

With the fix I posted above, it works with and without any file extensions.

**Comment by Wengier** [comment]
It turns out there is a bug in the DOS_OpenFile function resulting incorrect error code to be returned, causing this issue. It is fixed in PR #1550.
